cpp/Naive.cpp

```diff
@@ -1,65 +1,413 @@
+                                                  GM_Nonlinear_K(G_MIN, kx, ky) +
+                                                  dt / 2.0 * GM_Nonlinear_K_Loop(G_MIN, kx, ky);
+                });
+                DerivateNewMoment(G_MIN);
+
+                DxDy<fftw::mdbuffer<2u>> dGMinusPlus_Loop{X, Y};
+                for (int m = G_MIN + 1; m < LAST; ++m) {
+                    for_each_xy([&](Dim x, Dim y) {
+                        dGMinusPlus_Loop.DX(x, y) =
+                                std::sqrt(m) * dGM_Loop.DX(m - 1, x, y) + std::sqrt(m + 1) * dGM_Loop.DX(m + 1, x, y);
+                        dGMinusPlus_Loop.DY(x, y) =
+                                std::sqrt(m) * dGM_Loop.DY(m - 1, x, y) + std::sqrt(m + 1) * dGM_Loop.DY(m + 1, x, y);
+                    });
+
+                    auto bracketAParGMMinusPlus_K_Loop = halfBracket(sliceXY(dGM_Loop, A_PAR), dGMinusPlus_Loop);
+                    auto bracketPhiGM_K_Loop = halfBracket(dPhi_Loop, sliceXY(dGM_Loop, m));
+
+                    for_each_kxky([&](Dim kx, Dim ky) {
+                        GM_Nonlinear_K_Loop(m, kx, ky) = nonlinear::GM(m, bracketPhiGM_K_Loop(kx, ky),
+                                                                       bracketAParGMMinusPlus_K_Loop(kx, ky));
+                        // TODO(OPT) reuse star
+                        momentsNew_K(m, kx, ky) = exp_gm(m, hyper.nu_ei, dt) * moments_K(m, kx, ky) +
+                                                  dt / 2.0 * (1 + exp_gm(m, hyper.nu_ei, dt)) *
+                                                  GM_Nonlinear_K(m, kx, ky) +
+                                                  dt / 2.0 * GM_Nonlinear_K_Loop(m, kx, ky);
+                    });
+
+                    DerivateNewMoment(m);
                 }
+
+                // Compute G_{M-1}
+                auto bracketPhiGLast_K_Loop = halfBracket(dPhi, sliceXY(dGM, LAST));
+                auto bracketAParGLast_K_Loop = halfBracket(sliceXY(dGM, A_PAR), sliceXY(dGM, LAST));
+                for_each_kxky([&](Dim kx, Dim ky) {
+                    bracketAParGLast_K_Loop(kx, ky) *= nonlinear::GLastBracketFactor(M, kx, ky, hyper);
+                    bracketAParGLast_K_Loop(kx, ky) += rhoS / de * std::sqrt(LAST) * momentsNew_K(LAST - 1, kx, ky);
+                    // TODO Viriato adds this after the derivative
+                });
+
+                DxDy<fftw::mdbuffer<2u>> dBrLast_Loop{X, Y};
+                derivatives(bracketAParGLast_K_Loop, dBrLast_Loop);
+                auto bracketTotalGLast_K_Loop = halfBracket(sliceXY(dGM_Loop, A_PAR), dBrLast_Loop);
+
+                for_each_kxky([&](Dim kx, Dim ky) {
+                    GM_Nonlinear_K_Loop(LAST, kx, ky) = nonlinear::GLast(bracketPhiGLast_K_Loop(kx, ky),
+                                                                         bracketTotalGLast_K_Loop(kx, ky));
+                    // TODO(OPT) reuse star
+                    momentsNew_K(LAST, kx, ky) =
```

Comment:
Same comment as above applied here. Not quite right with the exp functions

Reply:
Fixed!